src/VisualStudio/Core/Def/Implementation/SemanticClassificationCache/SemanticClassificationCacheIncrementalAnalyzerProvider.cs

```diff
@@ -49,6 +62,12 @@ public override async Task AnalyzeDocumentAsync(Document document, SyntaxNode bo
                     return;
 
                 var solution = document.Project.Solution;
+
+                // Only cache classifications in non-LSP scenarios. LSP caching is handled separately.
```

Comment:
@dibarbet here as well...

interesting.  why is this @allisonchou ?  why not have the normal incremental analyzer handle this.  note: if this is the right way to go, then this comment needs beefing up.  Specifically, this is an example of a comment htat just repeats what the code is doing but doesn't provide clarity as to 'why'.  I cannot look at this code/comment and say: ah, yes, that makes sense, we def should be doing things in this fashion.

Reply:
The main issue with using the default logic is the method's call to [`document.IsOpen()`](https://github.com/dotnet/roslyn/blob/498d5b1f6feace039534035a08aa864c0cd26d1e/src/VisualStudio/Core/Def/Implementation/SemanticClassificationCache/SemanticClassificationCacheIncrementalAnalyzerProvider.cs#L48). This call is problematic in LSP for two reasons:

1) It bases whether the document is open on the non-LSP workspace rather than the LSP workspace.
2) Razor back-end generated documents won't be considered open even if the front-end document is open. This means we would have to special case Razor (via exposing a new document service, workspace service or something else).

Putting the caching logic in the LSP handler addresses both 1) and 2) by ensuring we only cache open files in the LSP workspace, since we only get semantic token requests for the file(s) currently open. In an ideal world, I think we'd have the `IsOpen` check correspond to the LSP workspace and that workspace can then specially check for open Razor documents, but I believe that is a much bigger work item.